[5기 - 최정은] 1~2주차 과제: 도서 관리 어플리케이션 제출합니다.

JeongeunChoi opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
> 콘솔을 통해 클라이언트로 부터 입력을 받아, 처리 후, 결과를 출력하는 도서 관리 애플리케이션입니다.

파일에 데이터를 저장하는 일반 모드와 메모리에 저장하는 테스트 모드 중 하나를 택해 동작하게 되며,
도서 등록, 전체 도서 목록 조회, 제목으로 도서 검색, 도서 대여, 도서 반납, 도서 분실, 도서 삭제 기능을 제공합니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
<img src="https://github.com/prgrms-be-devcourse/java-library-management/assets/77786996/55b43f94-4f90-4223-9186-faccde19b50e" width="350" height="250" />


### 디렉토리 구조 및 설명
<img src="https://github.com/prgrms-be-devcourse/java-library-management/assets/77786996/cfa37063-2590-4b42-b203-4547c8f70db9" width="350" height="450" />

Client.java
- 콘솔 입출력을 담당합니다.

domain
- Book.java
  - 도서에 관한 정보를 가지는 클래스
- BookStatus.java
  - 도서의 상태를 나타내는 enum

service
- LibraryManagemnetService.java
  - 클라이언트가 선택한 기능을 처리하는 비즈니스 로직
   - 요청한 기능에 대해 처리 후, Optional<ExceptionCode> 형태로 반환한다. 
   - 클라이언트에서는 Optional 객체가 비어있는 경우 정상 처리됨을 알리고, ExceptionCode가 존재하는 경우 메시지를 출력하여 알린다.

repository
- Repository.java
  - 일반 모드와 테스트 모드에서 제공하는 공통적인 기능을 명시하는 인터페이스
- FileRepository.java
  - 일반 모드에서 csv 파일에 대하여 도서 생성, 읽기, 갱신, 삭제 관련 기능을 수행
- InMemoryRepository.java
  - 테스트 모드에서 메모리에 저장된 정보에 대하여 도서 생성, 읽기, 갱신, 삭제 관련 기능을 수행

exception
- ExceptionCode.java
  - 도서 관리 기능 처리 중 발생할 수 있는 예외 코드와 메시지를 나타내는 enum

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
1. 도서 엔티티의 속성 중 책 반납 시간의 경우, 반납 시에 반납 시간에 대해 값을 설정하고 나머지 경우에는 null 값이 들어가도록 되어 있습니다. 이렇게 null 값을 넣도록 구현하여도 상관없나요? 또는 다른 방법이 있는지 궁금합니다!

2. 예외 처리의 경우, 요청한 기능에 대해 처리 후, Optional<ExceptionCode> 형태로 반환하여 예외가 있으면 콘솔에 출력해주도록 처리하고 있습니다. 좀 더 효율적이거나, 객체지향적인 방법이 있는지 궁금합니다!

3. 현재 테스트 코드를, FileRepository, InMemoryRepository에 대해 따로 작성한 상태입니다. 하지만 repository 변수에 대입되는 인스턴스를 제외하고는 코드가 동일합니다. 
이런 경우에는, 테스트 코드 파일을 현재처럼 따로 만드는게 좋을까요? 아니면 하나의 파일에서 테스트 하는게 좋을까요?
만약 후자라면, 어떤 식으로 코드를 구성해야 하는지도 궁금합니다!

4. static 메서드를 몇개 정의해서 사용해봤는데, static 사용은 객체지향적으로 구성 시에는 거의 사용하지 않는 편인가요? BookStatus와 ExceptionCode에서 값을 가져올 때 static을 사용하고 있습니다.(모두 enum 클래스입니다!)

5. 서비스 클래스에 대해서 테스트 코드를 작성하며, 의존 주입으로 레포지토리 또한 함께 사용되었는데 이렇게 작성하는 것이 맞는지 궁금합니다!

6. 클래스를 상세하게 나누기 보다는 사용 용도가 비슷한 경우 하나로 구현하였는데, 분리할 필요가 있는 부분이 있을까요?

7. 자바 강의를 통해, 디자인 패턴에 학습하긴 하였으나 코드에 적용하기에는 잘 떠오르지 않더라구요! 혹시 현재 코드에서 적용하면 좋을 디자인 패턴이 있는지도 궁금합니다!

이외에도 객체지향적인 부분이나 다른 부분에 대해서 고칠 부분이 있으면 따끔한 피드백 부탁드립니다!🙇‍♀️

JeongeunChoi committed:
chore: 프로젝트 초기 설정

JeongeunChoi committed:
feat: BookEntity, BookStatus 구현

JeongeunChoi committed:
feat: Repository 인터페이스 구현

JeongeunChoi committed:
feat: 일반 모드용 FileRepository 구현

JeongeunChoi committed:
feat: 테스트 모드용 InMemoryRepository 구현

JeongeunChoi committed:
feat: 예외코드 관리용 ExceptionCode 구현

JeongeunChoi committed:
feat: 도서 관리 비즈니스 로직 처리를 위한 LibraryManagementService 구현

JeongeunChoi committed:
feat: 콘솔 입출력을 위한 Client 구현

JeongeunChoi committed:
test: 일반 모드에서의 도서 관리 기능 테스트 코드 구현

JeongeunChoi committed:
test: 테스트 모드에서의 도서 관리 기능 테스트 코드 구현

JeongeunChoi committed:
chore: 도서 정보 저장을 위한 csv 파일 추가

hoon-space commented on src/main/java/org/example/Client.java:
```diff
@@ -0,0 +1,147 @@
+import java.util.Scanner;
+
+public class Client {
+    LibraryManagementService libraryManagementService;
+    Scanner scan = new Scanner(System.in);
+    private Integer nextBookId;
+    private final String bookInfoCsvPath = "data/bookInfo.csv";
```
질문) private 으로 선언이 된 것과 아닌 것의 기준을 알고싶어요

hoon-space commented on src/main/java/org/example/Client.java:
```diff
@@ -0,0 +1,147 @@
+        nextBookId = libraryManagementService.getNextBookId();
+    }
+
+    private void selectLibraryFunction() {
+        System.out.println("Q. 사용할 기능을 선택해주세요.\n1. 도서 등록\n2. 전체 도서 목록 조회\n3. 제목으로 도서 검색\n4. 도서 대여\n5. 도서 반납\n6. 도서 분실\n7. 도서 삭제");
```
JDK 17 이상 사용하신다면 """ 를 사용해도 좋을 것 같아요

hoon-space replied:
JDK 17 이상 사용하신다면 """ 를 사용해도 좋을 것 같아요

hoon-space commented on build.gradle:
```diff
@@ -0,0 +1,19 @@
+plugins {
+    id 'java'
+}
+
+group 'org.example'
+version '1.0-SNAPSHOT'
+
+repositories {
+    mavenCentral()
+}
+
```
JDK 버전은 어떤걸 사용하시나요 ?

hoon-space commented on src/main/java/org/example/exception/ExceptionCode.java:
```diff
@@ -0,0 +1,36 @@
+
+import org.example.domain.BookStatus;
+
+public enum ExceptionCode {
+    AVAILABLE_FOR_BORROW("[System] 원래 대여가 가능한 도서입니다."),
+    ALREADY_BORROWED("[System] 이미 대여 중인 도서입니다."),
+    BEING_ORGANIZED("[System] 정리 중인 도서입니다."),
+    LOST("[System] 분실 처리된 도서입니다."),
+    INVALID_BOOK("[System] 존재하지 않는 도서번호 입니다.");
```
👍

hoon-space replied:
👍

hoon-space commented on build.gradle:
```diff
@@ -0,0 +1,19 @@
+plugins {
+    id 'java'
+}
+
+group 'org.example'
+version '1.0-SNAPSHOT'
+
+repositories {
+    mavenCentral()
+}
+
```
JDK 버전은 어떤걸 사용하시나요 ?

JeongeunChoi replied:
JDK 17 사용하고 있습니다!

hoon-space replied:
JDK 버전을 명시해주는게 좋을 것 같아요. 
다른 곳에서 실행시킬때, 버전 명시가 되어 있지않으면 로컬 버전과 다르면 어떤 버전을 사용하는지 혼란스러울 것 같아요. 

=> 코드는 용상님이나 경민님꺼 참고하면 좋을 것 같아요

hoon-space commented on src/main/java/org/example/Client.java:
```diff
@@ -0,0 +1,147 @@
+
+import java.util.Scanner;
+
+public class Client {
+    LibraryManagementService libraryManagementService;
+    Scanner scan = new Scanner(System.in);
+    private Integer nextBookId;
+    private final String bookInfoCsvPath = "data/bookInfo.csv";
```
질문) private 으로 선언이 된 것과 아닌 것의 기준을 알고싶어요

JeongeunChoi replied:
엇 private로 모두 수정하겠습니다..! 감사합니다🙇‍♀️

hoon-space commented on src/main/java/org/example/domain/Book.java:
```diff
@@ -0,0 +1,94 @@
+    private String author;
+    private Integer pageSize;
+    private BookStatus status;
+    private Instant returnTime;
```
질문) Instant 타입을 사용하신 이유를 알 수 있을까요 ?? 
보통은 눈으로 읽을 수 있는 LocalDate 나 LocalDateTime 을 사용합니다.

JeongeunChoi replied:
오.. 저는 특별한 이유 없이 Instant를 사용해봤는데, 가시성을 고려하면 변경하는게 좋군요! 수정하도록 하겠습니다! 감사합니다!!👍

hoon-space commented on src/main/java/org/example/domain/Book.java:
```diff
@@ -0,0 +1,94 @@
+    private String author;
+    private Integer pageSize;
+    private BookStatus status;
+    private Instant returnTime;
```
추가로 반납이 되고 다시 빌리고 반납하면 이전에 저장을 했던 시간 값이 사라질텐데, 히스토리를 관리하는 객체를 하나 더 만들면 어떨까 싶습니다.

hoon-space commented on src/main/java/org/example/domain/BookStatus.java:
```diff
@@ -0,0 +1,28 @@
+package org.example.domain;
+
+public enum BookStatus {
```
보통 Enum 은 Type 을 명시하는 경우에 많이 사용해서, BookStatusType 으로 이름 변경해도 좋을 것 같아요

hoon-space commented on src/main/java/org/example/domain/BookStatus.java:
```diff
@@ -0,0 +1,28 @@
+    ORGANIZING("도서 정리중"),
+    LOST("분실됨");
+
+    private String koreanName;
```
name 이나 message 정도로 줄여도 좋을 것 같아요

hoon-space commented on src/main/java/org/example/domain/BookStatus.java:
```diff
@@ -0,0 +1,28 @@
+            }
+        }
+
+        return null;
```
저는 보통 Enum 값을 못찾은 경우에 예외를 날리는 편이에요. 
null 로 던지면 호출하는 쪽에서 어차피 예외처리를 해줘야해서 ...

hoon-space commented on src/main/java/org/example/domain/Book.java:
```diff
@@ -0,0 +1,94 @@
+    private BookStatus status;
+    private Instant returnTime;
+
+    public Book(Integer id, String title, String author, Integer pageSize) {
+        this.id = id;
+        this.title = title;
+        this.author = author;
+        this.pageSize = pageSize;
+        this.status = BookStatus.BORROW_AVAILABE;
+        this.returnTime = null;
+    }
+
+    public Book(Integer id, String title, String author, Integer pageSize, BookStatus status) {
+        this.id = id;
+        this.title = title;
+        this.author = author;
+        this.pageSize = pageSize;
+        this.status = status;
+        this.returnTime = null;
+    }
+
+    public Book(Integer id, String title, String author, Integer pageSize, BookStatus status, Instant returnTime) {
+        this.id = id;
+        this.title = title;
+        this.author = author;
+        this.pageSize = pageSize;
+        this.status = status;
+        this.returnTime = returnTime;
+    }
```
생성자를 직접 생성하는 것도 좋지만, 정적팩토리메서드 사용해봐도 좋을 것 같아요
https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/

hoon-space commented on src/main/java/org/example/domain/Book.java:
```diff
@@ -0,0 +1,94 @@
+
+    public boolean canBorrow() {
+        return status == BookStatus.BORROW_AVAILABE ||
+                (status == BookStatus.ORGANIZING && Duration.between(returnTime, Instant.now()).toMinutes() >= 5);
```
5 라는 숫자는 하드 코딩보다는 변수명을 부여하는게 좀 더 가독성 있을 것 같아요

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    // 전체 도서 목록 조회
+    public List<Book> searchBooks() {
+        return repository.findAllBooks();
```
전체니까 searchAllBooks ??

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -0,0 +1,84 @@
+        return repository.findAllBooks();
+    }
+
+    // 제목으로 도서 검색
+    public List<Book> searchBookByTitle(String title) {
+        return repository.findBookByTitle(title);
+    }
```
searchBookBy(String title) ?? 로 작성해도 좋을 것 같아요

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    // 도서 대여
+    public Optional<ExceptionCode> borrowBook(Integer bookId) {
+        Optional<Book> bookToBorrow = repository.findBookById(bookId);
+        return bookToBorrow.map(book -> {
+            if (book.canBorrow()) {
+                repository.updateBookStatus(bookId, BookStatus.BORROWING);
+                return Optional.<ExceptionCode>empty();
+            } else {
+                return Optional.of(ExceptionCode.getException(book.getStatus()));
+            }
+        }).orElse(Optional.of(ExceptionCode.INVALID_BOOK));
```
성공일때도 예외코드를 던지나요 ???

JeongeunChoi replied:
성공 시에는 비어있는 Optional을 반환하도록 하였습니다!
클라이언트 측에서, Optional에 예외코드가 들어가 있는 경우에만 예외를 출력해줍니다!

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -0,0 +1,84 @@
+        }).orElse(Optional.of(ExceptionCode.INVALID_BOOK));
+    }
+
+    public Integer getNextBookId() {
+        return repository.getNextBookId();
+    }
```
도서 ID 채번하는 로직이 맞을까요 ??
맞다면 레포지토리에 있어야할 것 같아요

hoon-space commented on src/main/java/org/example/repository/InMemoryRepository.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.stream.Collectors;
+
+public class InMemoryRepository implements Repository {
+    List<Book> books = new ArrayList<>();
```
질문) ArrayList 자료형을 사용하신 이유를 알 수 있을까요 ??

JeongeunChoi replied:
Book id를 통한 조회가 이루어지는 경우가 많으니, Map을 사용하는 편이 맞겠네요..! 수정해두도록 하겠습니다😀

hoon-space commented on src/main/java/org/example/Client.java:
```diff
@@ -0,0 +1,147 @@
+        int functionNumber = scan.nextInt();
+        scan.nextLine();
+
+        switch (functionNumber) {
+            case 1:
+                registerBook();
+                break;
+            case 2:
+                searchAllBooks();
+                break;
+            case 3:
+                searchBooksByTitle();
+                break;
+            case 4:
+                borrowBook();
+                break;
+            case 5:
+                returnBook();
+                break;
+            case 6:
+                lostBook();
+                break;
+            case 7:
+                deleteBook();
+                break;
+            default:
+                System.out.println("1부터 7까지의 정수 중 하나를 입력해주세요.\n");
+                break;
+        }
```
하드코딩된 숫자가 굉장히 많아서, Enum 으로 커맨드 관리하면 좋을 것 같아요.

hoon-space reviewed (commented):
첫 과제 고생 많으셨습니다. 👍

JeongeunChoi committed:
refactor: JDK 17의 텍스트블럭 사용

JeongeunChoi committed:
refactor: 접근제어자 private 누락된 부분 추가

JeongeunChoi committed:
refactor: Book, BookStatus 수정

- Book의 returnTime 속성 타입 LocalDateTime으로 수정
- BookStatus 이름 BookStatusType으로 이름 수정
- BookStatus 값을 찾지 못한 경우 null이 아닌 예외를 반환하도록 수정

JeongeunChoi committed:
refactor: 서비스 클래스의 전체 도서 조회 및 제목으로 도서 검색 함수명 수정

JeongeunChoi committed:
refactor: 도서 기능 선택 시, enum으로 커맨드 관리하도록 수정

JeongeunChoi committed:
refactor: 예외를 서비스에서 던지고, 클라이언트에서 받아서 처리하도록 수정

hoon-space commented on src/main/java/org/example/Client.java:
```diff
@@ -68,8 +79,9 @@ private void selectLibraryFunction() {
     private void registerBook() {
-        System.out.println("[System] 도서 등록 메뉴로 넘어갑니다.");
-        System.out.println("Q. 등록할 도서 제목을 입력하세요.");
+        System.out.println("""
+                [System] 도서 등록 메뉴로 넘어갑니다.
+                Q. 등록할 도서 제목을 입력하세요.""");
```
NIT)
### AS-IS
```java
        System.out.println("""
                [System] 도서 등록 메뉴로 넘어갑니다.
                Q. 등록할 도서 제목을 입력하세요.""");
```

### TO-BE
```java
        System.out.println("""
                [System] 도서 등록 메뉴로 넘어갑니다.
                Q. 등록할 도서 제목을 입력하세요."""
          );
```

hoon-space reviewed (approved):
첫 과제 고생 많으셨습니다.
어풉드릴게요 🙇‍♂️

JeongeunChoi committed:
chore: JDK 17 버전 명시

JeongeunChoi committed:
style: 텍스트 블럭 코드 포맷 수정

JeongeunChoi committed:
refactor: Book 인스턴스를 담는 List를 Map으로 수정

JeongeunChoi committed:
refactor: 도서 반납 5분 후 대여 가능으로 변경하도록 기능 수정

JeongeunChoi committed:
refactor: Book에 정적 팩토리 메서드 적용

JeongeunChoi committed:
refactor: 클라이언트로부터 콘솔 입출력 시에 사용되는 디렉토리 생성

ordilov reviewed (approved):
과제 고생하셨습니다~
우선 테스트 꼼꼼하게 작성해주신 부분이 되게 좋았습니다. 👍 
그리고 훈님도 답변해주시겠지만 질문주신거 제 의견드리겠습니다!

1번) 질문 답변드리려고 했는데 `returnTime`? 이 사라졌는지 모델에서 안보이네요 😢 

2번) 취향차이라 해당 부분에 답글 달았습니다!

3번 + 5번) 테스트의 경우 저라면 하나의 파일에서 테스트합니다.
서비스 테스트라면 의존하는 레포지토리 관련없이 서비스 테스트 하나로 작성합니다.
서비스가 의존하는건 인터페이스고, 테스트할 때 구현체는 어떤 걸 쓰셔도 무방합니다.
다만 이번에 구성해주신 것처럼 각 레포지토리에 따라 결과값이 잘나오나 궁금하실 수 있는데 그 부분은 서비스 테스트가 아니라 각 레포지토리에 대한 테스트를 작성해주시면 됩니다.
정리하면 의존하고 있는 객체에 따라 테스트를 분리하기보다 의존하는 각 객체에 대한 테스트를 짜기입니다!

JeongeunChoi committed:
style: 서비스 클래스에서 orElseThrow() 사용으로 ifPresent + if 중복 제거

JeongeunChoi committed:
refactor: 서비스의 nextBookId 함수를 레포지토리로 이동

JeongeunChoi commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -0,0 +1,94 @@
+
+    // 도서 대여
+    public void borrowBook(Integer bookId) {
+        Optional<Book> bookToBorrow = repository.findBookById(bookId);
```
orElseThrow 적용하니 훨씬 간단해지네요! 감사합니다😀

JeongeunChoi commented on src/main/java/org/example/domain/Book.java:
```diff
@@ -0,0 +1,70 @@
+    private Integer id;
+    private String title;
+    private String author;
+    private Integer pageSize;
```
헉 생각해보지 못한 부분인데 Integer보다는 int를 더 많이 사용되는 편인가요?👀

추가로
> 1번) 질문 답변드리려고 했는데 returnTime? 이 사라졌는지 모델에서 안보이네요 😢

이부분은 해당 속성이 반납 5분 후 책 상태 변경을 위해 사용되었는데, 스케줄러를 이용하면서 삭제되었습니다! 꼼꼼히 봐주셔서 감사합니다 :)

hoon-space replied:
대신 답변드리자면, NULL 허용 여부에 따라 달라지는 것 같아요. (물론 박싱 언박싱도 있겠지만 이 이야기는 아닐테니 ....)
보통 pageSize 는 null 일수는 없으니 int 를 사용하라고 하신 것 같습니다.

관련한 글: https://blog.naver.com/PostView.naver?blogId=ssookie_&logNo=222432977225&parentCategoryNo=&categoryNo=6&viewDate=&isShowPopularPosts=true&from=search

JeongeunChoi replied:
아하 속성의 null 여부에 따라 int와 Integer를 고민해봐야하는군요... 답변 감사합니다!

JeongeunChoi committed:
refactor: 콘솔 입력 유효성 검사 추가

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -36,56 +35,44 @@ public List<Book> searchBookBy(String title) {
-            }
-        });
-        bookToBorrow.orElseThrow(() -> new IllegalArgumentException(ExceptionCode.INVALID_BOOK.getMessage()));
+        Book bookToBorrow = repository.findBookById(bookId).orElseThrow(() -> new IllegalArgumentException(ExceptionCode.INVALID_BOOK.getMessage()));
+        if (bookToBorrow.canBorrow()) {
```
NIT) 단순히 책 ID로 조회한거면 변수명을  bookToBorrow -> book 으로 해도 좋을 것 같아요

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -36,56 +35,44 @@ public List<Book> searchBookBy(String title) {
+        Book bookToReturn = repository.findBookById(bookId).orElseThrow(() -> new IllegalArgumentException(ExceptionCode.INVALID_BOOK.getMessage()));
+        if (bookToReturn.canReturn()) {
+            repository.updateBookStatus(bookId, BookStatusType.ORGANIZING);
+            ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+            scheduler.schedule(() -> {
+                repository.updateBookStatus(bookId, BookStatusType.BORROW_AVAILABE);
+                scheduler.shutdown();
+            }, organizingMinutes, TimeUnit.MINUTES);
```
해당 메서드 요청을 할 떄 마다 newSingleThreadScheduledExecutor() 호출로 스레드 하나가 더 생길 것 같아요.
이번 과제는 콘솔이니 넘어가도 좋지만, 나중에 웹 서버를 개발할 때, 스레드 관리가 정말 중요합니다. 🥲

JeongeunChoi replied:
스레드 생성하지 않고 상태를 관리하는 방법으로 구현하려고 하니, 책 상태를 확인하는 곳마다 반납 시간 점검을 해주어야 해서 사용해보았는데 오히려 스레드 사용이 더 문제가 될 수 있군요! 다음 과제부터는 주의해서 사용하도록 하겠습니다! 코멘트 감사합니다😀

hoon-space commented on src/main/java/org/example/service/LibraryManagementService.java:
```diff
@@ -36,56 +35,44 @@ public List<Book> searchBookBy(String title) {
-            }
-        });
-        lostBook.orElseThrow(() -> new IllegalArgumentException(ExceptionCode.INVALID_BOOK.getMessage()));
+        Book lostBook = repository.findBookById(bookId).orElseThrow(() -> new IllegalArgumentException(ExceptionCode.INVALID_BOOK.getMessage()));
```
NIT) 뭔가 네이밍만 보았을 때는 이미 lost 된 책 같은 ....

hoon-space commented on src/test/java/org/example/service/ServiceWithFileRepositoryTest.java:
```diff
@@ -164,7 +160,7 @@ void returnBook_Fail_BorrrowAvailable() {
     void lostBook_Success() {
         //given
-        libraryManagementService.registerBook(createTestBook());
+        repository.saveBook(createTestBook());
```
NIT) 뒤에 인자가 Book 이니 save(Book book) 으로 네이밍해도 좋을 것 같아요.

hoon-space reviewed (approved):
첫 과제 리뷰 반영해주시느라 고생 많으셨습니다. 🙇‍♂️

JeongeunChoi committed:
style: 변수명 및 함수명 수정
